backend/internal/core/distributed_locks/redis_locker.go

```diff
@@ -0,0 +1,135 @@
+package distributedlocks
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+)
+
+type RedisLocker struct {
+	client      *redis.Client
+	lockTimeout time.Duration
+}
+
+// NewRedisLocker creates a new RedisLocker instance.
+func NewRedisLocker(client *redis.Client, lockTimeout time.Duration) *RedisLocker {
+	return &RedisLocker{
+		client:      client,
+		lockTimeout: lockTimeout,
+	}
+}
+
+// AcquireNodesLocks acquires locks for the given node IDs.
+func (l *RedisLocker) AcquireNodesLocks(ctx context.Context, nodeIDs []uint32) error {
+	if err := l.acquireKeys(ctx, lockKeys(nodeIDs, nodeLockKey)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AcquireWorkflowLock acquires a lock for the given workflow ID.
+func (l *RedisLocker) AcquireWorkflowLock(ctx context.Context, nodeIDs []uint32, workflowID string) error {
+	keys := lockKeys(nodeIDs, func(id uint32) string {
+		return workflowLockKey(id, workflowID)
+	})
+
+	if err := l.acquireKeys(ctx, keys); err != nil {
+		//rollback nodes locks
+		nodeLockKeys := lockKeys(nodeIDs, nodeLockKey)
+		if rollErr := l.rollbackLocks(ctx, nodeLockKeys); rollErr != nil {
+			return rollErr
+		}
+		return err
+	}
+
+	return nil
+}
+
+func nodeLockKey(nodeID uint32) string {
+	return fmt.Sprintf("locked:%d", nodeID)
+}
+
+func workflowLockKey(nodeID uint32, workflowID string) string {
+	return fmt.Sprintf("used:%d:%s", nodeID, workflowID)
```

Comment:
extract used and locked prefixes to consts and reuse them everywhere e.g to 

`kubecloud:env:`